Objects/frameobject.c

```diff
@@ -13,9 +13,348 @@
 #include "pycore_frame.h"
 #include "opcode.h"               // EXTENDED_ARG
 
-
 #define OFF(x) offsetof(PyFrameObject, x)
 
+
+// Utilities
+
+// Returns borrowed reference or NULL
+static PyObject *
+framelocalproxy_getval(PyFrameObject *frame, PyCodeObject *co, int i)
+{
+    PyObject **fast = _PyFrame_GetLocalsArray(frame->f_frame);
+    _PyLocals_Kind kind = _PyLocals_GetKind(co->co_localspluskinds, i);
+
+    PyObject *value = fast[i];
+    PyObject *cell = NULL;
+
+    if (value == NULL) {
+        return NULL;
+    }
+
+    if (kind == CO_FAST_FREE || kind & CO_FAST_CELL) {
+        // The cell was set when the frame was created from
+        // the function's closure.
+        assert(PyCell_Check(value));
+        cell = value;
+    }
+
+    if (cell != NULL) {
+        value = PyCell_GET(cell);
+    }
+
+    if (value == NULL) {
+        return NULL;
+    }
+
+    return value;
+}
+
+// Type functions
+
+static PyObject *
+framelocalsproxy_keys(PyObject *self, PyObject *__unused)
+{
+    PyObject *names = PyList_New(0);
+    PyFrameObject *frame = ((PyFrameLocalsProxyObject*)self)->frame;
+    PyCodeObject *co = PyFrame_GetCode(frame);
+
+    for (int i = 0; i < co->co_nlocalsplus; i++) {
+        PyObject *val = framelocalproxy_getval(frame, co, i);
+        if (val) {
+            PyObject *name = PyTuple_GET_ITEM(co->co_localsplusnames, i);
+            PyList_Append(names, name);
+        }
+    }
+
+    // Iterate through the extra locals
+    Py_ssize_t i = 0;
+    PyObject *key = NULL;
+    PyObject *value = NULL;
+
+    if (frame->f_extra_locals) {
```

Comment:
Given the way `framelocalsproxy_new` is defined, I believe these various `frame->f_extra_locals` new checks in proxy methods can be safely converted to use `assert()`.

Alternatively, the creation of `f_extra_locals` could be made even lazier than it is at the moment. The current implementation creates `f_extra_locals` as soon as a proxy instance is created. However, most proxy instances are presumably either going to be used in a read-only fashion, or used to modify fast locals entries.

Moving the `f_extra_locals` creation to the first `framelocalsproxy_setitem` call that doesn't match a fast locals entry would delay the creation of the extra dict until the first moment it is genuinely needed (saving a value that doesn't have anywhere else to go).